mono/profiler/mono-profiler-log.c

```diff
@@ -4285,49 +4285,72 @@ new_filename (const char* filename)
 	return res;
 }
 
-static void*
-helper_thread (void* arg)
+static void
+add_to_fd_set (fd_set *set, int fd, int *max_fd)
+{
+	/*
+	 * This should only trigger for the basic FDs (server socket, pipe, perf
+	 * events) at startup if for some mysterious reason they're too large. In
+	 * this case, the profiler really can't function, and we're better off
+	 * printing an error and exiting.
+	 */
+	if (fd >= FD_SETSIZE) {
+		fprintf (stderr, "File descriptor is out of bounds for fd_set: %d\n", fd);
+		exit (1);
+	}
+
+	FD_SET (fd, set);
+
+	if (*max_fd < fd)
+		*max_fd = fd;
+}
+
+static void *
+helper_thread (void *arg)
 {
-	MonoProfiler* prof = (MonoProfiler *)arg;
-	int command_socket;
-	int len;
-	char buf [64];
+	MonoProfiler *prof = (MonoProfiler *) arg;
 
 	mono_threads_attach_tools_thread ();
 	mono_native_thread_set_name (mono_native_thread_id_get (), "Profiler helper");
 
 	MonoProfilerThread *thread = init_thread (prof, FALSE);
 
-	//fprintf (stderr, "Server listening\n");
-	command_socket = -1;
+	GArray *command_sockets = g_array_new (FALSE, FALSE, sizeof (int));
+
 	while (1) {
 		fd_set rfds;
-		struct timeval tv;
 		int max_fd = -1;
+
 		FD_ZERO (&rfds);
-		FD_SET (prof->server_socket, &rfds);
-		max_fd = prof->server_socket;
-		FD_SET (prof->pipes [0], &rfds);
-		if (max_fd < prof->pipes [0])
-			max_fd = prof->pipes [0];
-		if (command_socket >= 0) {
-			FD_SET (command_socket, &rfds);
-			if (max_fd < command_socket)
-				max_fd = command_socket;
-		}
+
+		add_to_fd_set (&rfds, prof->server_socket, &max_fd);
+		add_to_fd_set (&rfds, prof->pipes [0], &max_fd);
+
+		for (gint i = 0; i < command_sockets->len; i++)
+			add_to_fd_set (&rfds, g_array_index (command_sockets, int, i), &max_fd);
+
 #if USE_PERF_EVENTS
 		if (perf_data) {
-			int i;
-			for ( i = 0; i < num_perf; ++i) {
+			for (int i = 0; i < num_perf; ++i) {
 				if (perf_data [i].perf_fd < 0)
 					continue;
-				FD_SET (perf_data [i].perf_fd, &rfds);
-				if (max_fd < perf_data [i].perf_fd)
-					max_fd = perf_data [i].perf_fd;
+
+				add_to_fd_set (&rfds, perf_data [i].perf_fd, &max_fd);
 			}
 		}
 #endif
 
+		struct timeval tv = { .tv_sec = 1, .tv_usec = 0 };
+
+		// Sleep for 1sec or until a file descriptor has data.
+		if (select (max_fd + 1, &rfds, NULL, NULL, &tv) == -1) {
+			if (errno == EINTR)
+				continue;
+
+			fprintf (stderr, "Error in mono-profiler-log server: %s", strerror (errno));
+			exit (1);
```

Comment:
`select` can return these errors: `EINTR` (signal, we handle this one), `EBADF` (bad FD, indicates a programming error, because the server socket and pipes should not be closed at this point), `EINVAL` (bad arguments, programming error), `ENOMEM` (OOM, nothing we can do at that point).